Adding NET 4.8 WinForms Accessibility changes compatibility document. (#969)
Microsoft/dotnet: Documentation/compatibility/winforms-accessibility-changes-48.md

## Accessibility improvements in Windows Forms controls for .NET 4.8

### Scope
Major

### Version Introduced
4.8

### Source Analyzer Status
NotPlanned

### Change Description

> Windows Forms Framework is improving how it works with accessibility technologies to better support Windows Forms customers. These include the following changes:

Reviewer: ```suggestion
The Windows Forms Framework is continuing to improve how it works with accessibility technologies to better support Windows Forms customers. These include the following changes:
```

- Changes to improve display during High Contrast mode.
- Changes to interaction with Narrator.

> - Changes in Accessible hierarchy. (improving navigation through UI Automation tree)

Reviewer: ```suggestion
- Changes in the Accessible hierarchy (improving navigation through the UI Automation tree).
```

- [x] Quirked
- [ ] Build-time break

### Recommended Action
__How to opt in or out of these changes__
  

> In order for the application to benefit from these changes, it must run on the .NET Framework 4.8. The application can benefit from these changes in either of the following ways:

Reviewer 2: ```suggestion
In order for the application to benefit from these changes, it must run on the .NET Framework 4.8. The application can opt-in into these changes in either of the following ways:
```

- It is recompiled to target the .NET Framework 4.8. These accessibility changes are enabled by default on Windows Forms applications that target the .NET Framework 4.8.

> - It targets the .NET Framework 4.7.2 or earlier version and opts out of the legacy accessibility behaviors by adding the following [AppContext Switch](https://docs.microsoft.com/dotnet/framework/configure-apps/file-schema/runtime/appcontextswitchoverrides-element) to the `<runtime>` section of the app config file and setting it to `false`, as the following example shows.

Reviewer: ```suggestion
- It targets the .NET Framework 4.7.2 or earlier version and opts out of the legacy accessibility behaviors by adding the following [AppContext switch](https://docs.microsoft.com/dotnet/framework/configure-apps/file-schema/runtime/appcontextswitchoverrides-element) to the `<runtime>` section of the app config file and setting it to `false`, as the following example shows.
```

      <startup>
        <supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.7"/>
      </startup>
      <runtime>
        <!-- AppContextSwitchOverrides value attribute is in the form of 'key1=true|false;key2=true|false  -->
        <AppContextSwitchOverrides value="Switch.UseLegacyAccessibilityFeatures=false;Switch.UseLegacyAccessibilityFeatures.2=false;Switch.UseLegacyAccessibilityFeatures.3=false" />
      </runtime>
    </configuration>
```    
Note that to opt in to the accessibility features added in .NET Framework 4.8, you must also opt in to accessibility features of .NET Framework 4.7.1 and 4.7.2 as well.
Applications that target the .NET Framework 4.8 and want to preserve the legacy accessibility behavior can opt in to the use of legacy accessibility features by explicitly setting this AppContext switch to `true`.

__Use of OS-defined colors in High Contrast themes__
// To be filled.

__Improved Narrator support__
- Narrator now announces the sort direction of the <xref:System.Windows.Forms.DataGridViewColumn> when announcing an accessible name of a <xref:System.Windows.Forms.DataGridViewCell>. 

__Improved Checked ListBox Accessibility support__

> - Improved Narrator support for <xref:System.Windows.Forms.CheckedListBox> control. At this moment when navigating to the <xref:System.Windows.Forms.CheckedListBox> control using keyboard, Narrator focuses the <xref:System.Windows.Forms.CheckedListBox> item and announces it.

Reviewer: ```suggestion
- Improved Narrator support for the <xref:System.Windows.Forms.CheckedListBox> control. When navigating to the <xref:System.Windows.Forms.CheckedListBox> control using the keyboard, Narrator focuses the <xref:System.Windows.Forms.CheckedListBox> item and announces it.
```

__Improved ComboBox Accessibility support__

> - Enabled UI Automation support for <xref:System.Windows.Forms.ComboBox> control with ability to use UI Automation notifications and other UI Automation features.

Reviewer: ```suggestion
- Enabled UI Automation support for the <xref:System.Windows.Forms.ComboBox> control, with the ability to use UI Automation notifications and other UI Automation features.
```

__Improved DataGridView Accessibility support__

> - Enabled UI Automation support for <xref:System.Windows.Forms.DataGridView> control with ability to use UI Automation notifications and other UI Automation features.

Reviewer 2: ```suggestion
- Enabled UI Automation support for <xref:System.Windows.Forms.DataGridView> control with ability to use UI Automation notifications, other UI Automation features, and keyboard navigation of the control.
```
Though I'm not 100% sure, please verify, I think we added f3 for sort order access

Author: I looked at this and found out that F3-sorting was added in 4.7.2.
.

> - The UI Automation element which corresponds to the <xref:System.Windows.Forms.DataGridViewComboBoxEditingControl> or <xref:System.Windows.Forms.DataGridViewTextBoxEditingControl> is now a child of corresponding editing cell.

Reviewer: ```suggestion
- The UI Automation element which corresponds to the <xref:System.Windows.Forms.DataGridViewComboBoxEditingControl> or <xref:System.Windows.Forms.DataGridViewTextBoxEditingControl> is now a child of the corresponding editing cell.
```

> __Improved Link Label Accessibility support__

Reviewer 2: ```suggestion
__Improved LinkLabel Accessibility support__
```

Author: Thanks!

> - Improved <xref:System.Windows.Forms.LinkLabel> control accessibility: Narrator announces the disabled state for the link if corresponding <xref:System.Windows.Forms.LinkLabel> control is disabled.

Reviewer: ```suggestion
- Improved <xref:System.Windows.Forms.LinkLabel> control accessibility: Narrator announces the disabled state for the link if the corresponding <xref:System.Windows.Forms.LinkLabel> control is disabled.
```

> __Improved Progress Bar Accessibility support__

Reviewer 2: ```suggestion
__Improved ProgressBar Accessibility support__
```

> - Enabled UI Automation support for <xref:System.Windows.Forms.ProgressBar> control with ability to use UI Automation notifications and other UI Automation features. Developers are now able to use UI Automation notifications which Narrator can announce to indicate the progress.

Reviewer 2: @M-Lipin  - is there a link to UIA notification you could add here?

Author: Thanks for the comment!
I've added the link in 15c3da6c5585bdd314b3b75a285626a016c5d011.

> - Enabled UI Automation support for <xref:System.Windows.Forms.ProgressBar> control with ability to use UI Automation notifications and other UI Automation features. Developers are now able to use UI Automation notifications which Narrator can announce to indicate the progress.

Reviewer: ```suggestion
- Enabled UI Automation support for the <xref:System.Windows.Forms.ProgressBar> control with the ability to use UI Automation notifications and other UI Automation features. Developers are now able to use UI Automation notifications which Narrator can announce to indicate progress.
```

> __Improved Property Grid Accessibility support__

Reviewer 2: ```suggestion
__Improved PropertyGrid Accessibility support__
```

> - Enabled UI Automation support for <xref:System.Windows.Forms.PropertyGrid> control with ability to use UI Automation notifications and other UI Automation features.

Reviewer: ```suggestion
- Enabled UI Automation support for the <xref:System.Windows.Forms.PropertyGrid> control, with the ability to use UI Automation notifications and other UI Automation features.
```

> - The UI Automation element which corresponds to the currently editing property is now a child of corresponding property item UI Automation element.

Reviewer 2: ```suggestion
- The UI Automation element which corresponds to the currently edited property is now a child of the corresponding property item UI Automation element.
```

Author: Thanks!

> - The UI Automation property item element is now a child of corresponding category if the parent <xref:System.Windows.Forms.PropertyGrid> control is set to Category view.

Reviewer 2: ```suggestion
- The UI Automation property item element is now a child of the corresponding category element if the parent <xref:System.Windows.Forms.PropertyGrid> control is set to category view.
```
I'm not sure about capitalization of the word `category`.

Author: Thanks. I agree, category should not have 'c' capitalized.

> __Improved Tool Strip support__

Reviewer 2: ```suggestion
__Improved ToolStrip support__
```

> - Enabled UI Automation support for <xref:System.Windows.Forms.ToolStrip> control with ability to use UI Automation notifications and other UI Automation features.

Reviewer: ```suggestion
- Enabled UI Automation support for the <xref:System.Windows.Forms.ToolStrip> control, with the ability to use UI Automation notifications and other UI Automation features.
```

- Improved navigation through <xref:System.Windows.Forms.ToolStrip> items. 

> - In items mode Narrator focus does not disappear and does not go to hidden items.

Reviewer: ```suggestion
- In items mode, Narrator focus does not disappear and does not go to hidden items.
```

__Improved Visual cues__

> - An empty <xref:System.Windows.Forms.CheckedListBox> control now displays a focus indicator when receives focus.

Reviewer 2: ```suggestion
- An empty <xref:System.Windows.Forms.CheckedListBox> control now displays a focus indicator when it receives focus.
```

```xml
    <?xml version="1.0" encoding="utf-8"?>
    <configuration>
      <startup>
        <supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.7"/>
      </startup>
      <runtime>
        <!-- AppContextSwitchOverrides value attribute is in the form of 'key1=true|false;key2=true|false  -->
        <AppContextSwitchOverrides value="Switch.UseLegacyAccessibilityFeatures=false;Switch.UseLegacyAccessibilityFeatures.2=false;Switch.UseLegacyAccessibilityFeatures.3=false" />
      </runtime>
    </configuration>
```    
Note that to opt in to the accessibility features added in .NET Framework 4.8, you must also opt in to accessibility features of .NET Framework 4.7.1 and 4.7.2 as well.
Applications that target the .NET Framework 4.8 and want to preserve the legacy accessibility behavior can opt in to the use of legacy accessibility features by explicitly setting this AppContext switch to `true`.

> 

Reviewer 3: ```suggestion
Enabling the keyboard ToolTip invocation support requires adding the `Switch.System.Windows.Forms.UseLegacyToolTipDisplay=false` line to the AppContextSwitchOverrides value:
`<AppContextSwitchOverrides value="Switch.UseLegacyAccessibilityFeatures=false;Switch.UseLegacyAccessibilityFeatures.2=false;Switch.UseLegacyAccessibilityFeatures.3=false" />`

Note that enabling this feature requires opting in to the aforementioned accessibility features of .NET Framework 4.7.1 - 4.8.

```

- Enabled UI Automation support for the <xref:System.Windows.Forms.PropertyGrid> control, with the ability to use UI Automation notifications and other UI Automation features.
- The UI Automation element which corresponds to the currently edited property is now a child of the corresponding property item UI Automation element.
- The UI Automation property item element is now a child of the corresponding category element if the parent <xref:System.Windows.Forms.PropertyGrid> control is set to category view.

__Improved ToolStrip support__
- Enabled UI Automation support for the <xref:System.Windows.Forms.ToolStrip> control, with the ability to use UI Automation notifications and other UI Automation features.
- Improved navigation through <xref:System.Windows.Forms.ToolStrip> items. 
- In items mode, Narrator focus does not disappear and does not go to hidden items.

__Improved Visual cues__
- An empty <xref:System.Windows.Forms.CheckedListBox> control now displays a focus indicator when it receives focus.

For an overview of UI automation, see the [UI Automation Overview](https://docs.microsoft.com/dotnet/framework/ui-automation/ui-automation-overview).</p>

> 

Reviewer 3: ```suggestion
__Invoking controls' ToolTips with a keyboard__
 Control tooltip can now be invoked by focusing the control with keyboard. 
 This feature needs to be enabled explicitly for the application (see section __"How to opt in or out of these changes"__)
 
```

### Affected APIs 

### Category
Windows Forms

<!--
    ### Original Bug
442899
526702
533226
657355
661319
642548
537224
549558
581351
560840
497307

> 

Reviewer 3: ```suggestion
548792
686499

```

```xml
    <?xml version="1.0" encoding="utf-8"?>
    <configuration>
      <startup>
        <supportedRuntime version="v4.0" sku=".NETFramework,Version=v4.7"/>
      </startup>
      <runtime>
        <!-- AppContextSwitchOverrides value attribute is in the form of 'key1=true|false;key2=true|false  -->
        <AppContextSwitchOverrides value="Switch.UseLegacyAccessibilityFeatures=false;Switch.UseLegacyAccessibilityFeatures.2=false;Switch.UseLegacyAccessibilityFeatures.3=false" />
      </runtime>
    </configuration>
```    
Note that to opt in to the accessibility features added in .NET Framework 4.8, you must also opt in to accessibility features of .NET Framework 4.7.1 and 4.7.2 as well.
Applications that target the .NET Framework 4.8 and want to preserve the legacy accessibility behavior can opt in to the use of legacy accessibility features by explicitly setting this AppContext switch to `true`.
Enabling the keyboard ToolTip invocation support requires adding the `Switch.System.Windows.Forms.UseLegacyToolTipDisplay=false` line to the AppContextSwitchOverrides value:

> `<AppContextSwitchOverrides value="Switch.UseLegacyAccessibilityFeatures=false;Switch.UseLegacyAccessibilityFeatures.2=false;Switch.UseLegacyAccessibilityFeatures.3=false" />`

Reviewer 2: ```suggestion
`<AppContextSwitchOverrides value="Switch.UseLegacyAccessibilityFeatures=false;Switch.UseLegacyAccessibilityFeatures.2=false;Switch.UseLegacyAccessibilityFeatures.3=false;Switch.System.Windows.Forms.UseLegacyToolTIpDisplay=false" />`
```

> Note that enabling this feature requires opting in to the aforementioned accessibility features of .NET Framework 4.7.1 - 4.8.

Reviewer 2: I would add that if any of the accessibility features are not opted in but the tooltip display feature is opted in, users would see a runtime exception on the first access to these features, most likely on application start up. Please specify exception type here.

Author: Thank you, I've added a note about enabling keyboard tooltip feature as you suggested. Thanks!
